### PR TITLE
Feature/add tests

### PR DESCRIPTION
**Pull request by ag-ramachandran**
# Required for all PRs

- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [ ] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #

### CODE REVIEW

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 1]
```diff
@@ -237,8 +242,23 @@ func init() {
 	})
 }
 
-func createRealIngestor(client localClient, database string, tableName string) (localIngestor, error) {
-	ingestor, err := ingest.New(client.(*kusto.Client), database, tableName, ingest.WithStaticBuffer(bufferSize, maxBuffers))
+// For each table create the ingestor
+func createIngestorByTable(client *kusto.Client, database string, tableName string, ingestionType string) (ingest.Ingestor, error) {
+	var ingestor ingest.Ingestor
+	var err error
+	if strings.ToLower(ingestionType) == managedIngestion {
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:250]: Maybe also add a constant for queued, so the users could specify either one, and anything else would be an error.

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 2]
```diff
@@ -237,8 +242,23 @@ func init() {
...
+		mi, err := ingest.NewManaged(client, database, tableName)
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:251]: We do not pass the static buffer to the queued ingest client inside of the managed client - how does that affect memory consumption?

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 3]
```diff
@@ -35,10 +35,10 @@ type AzureDataExplorer struct {
 	MetricsGrouping string          `toml:"metrics_grouping_type"`
 	TableName       string          `toml:"table_name"`
 	CreateTables    bool            `toml:"create_tables"`
-	client          localClient
-	ingesters       map[string]localIngestor
+	IngestionType   string          `toml:"ingestion_type"`
+	client          *kusto.Client
+	ingesters       map[string]ingest.Ingestor
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go]: Not yours - but change to "ingestors"Not yours - but change to "ingestors"

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 4]
```diff
@@ -18,217 +16,280 @@ import (
...
+			name:                      "Table per metric type",
+			inputMetric:               mockMetricsMulti,
+			createTables:              true,
+			client:                    fakeClient,
+			metricsGrouping:           tablePerMetric,
+			tableNameToExpectedResult: expectedResultMap2,
 		},
 	}
 
 	for _, tC := range testCases {
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer_test.go]: using the name "tC" was not yours, but now it's a bit much. Change this and all other 2 letter names to something good.

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 5]
```diff
@@ -18,217 +16,280 @@ import (
...
 		t.Run(tC.name, func(t *testing.T) {
 			serializer, err := telegrafJson.NewSerializer(time.Second, "", "")
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:83]: you might want to parallel tests here and elsewhere (t.Parallel at the start of the function, but don't forget to pin the test case - https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721)

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 6]
```diff
@@ -18,217 +16,280 @@ import (
...
-				},
-				"tags": map[string]interface{}{
-					"tag1": "value1",
-				},
-				"timestamp": float64(time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC).UnixNano() / int64(time.Second)),
-			},
+			name:                      "Valid metric",
+			inputMetric:               mockMetrics,
+			createTables:              true,
+			client:                    fakeClient,
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer_test.go]: If all of them accept fakeClient, it doesn't really need to be a parameter.

**Review by AsafMah — COMMENTED** [review]
Good job! some small comments

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go (hunk 2 above)
Comment (ag-ramachandran) [reply plugins/outputs/azure_data_explorer/azure_data_explorer.go:251]: Good question @AsafMah , we found the change was not much if we pass the static buffer. Do we want to pass the static buffer in either cases (managed and queued)

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 7]
```diff
@@ -18,217 +16,288 @@ import (
...
+			expectedError: "metrics grouping type is not valid",
+		},
+	}
 
-				if tC.metricsGrouping == singleTable {
-					expectedNameOfTable = tC.tableName
-					createdIngestor = plugin.ingesters[expectedNameOfTable]
-				}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:198]:
```suggestion
		testCase := testCase // capture to allow parallelism
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 8]
```diff
@@ -18,217 +16,288 @@ import (
...
+			createTables:              true,
+			metricsGrouping:           tablePerMetric,
+			tableNameToExpectedResult: expectedResultMap2,
 		},
 	}
 
-	for _, tC := range testCases {
-		t.Run(tC.name, func(t *testing.T) {
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
```
Comment (AsafMah) [inline plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:81]:
```suggestion
		testCase := testCase // capture to allow parallelism
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go (hunk 2 above)
Comment (AsafMah) [reply plugins/outputs/azure_data_explorer/azure_data_explorer.go:251]: We'll keep it like this for now.

**Review by AsafMah — APPROVED** [review]

**Review by asaharn — APPROVED** [review]

**Pull request merged**